src/glib/mine/svm.h

```diff
@@ -13,376 +13,640 @@
-    // ignored by the learning when kernel_type!=polynomial
-    svm_parameter.degree = 0;
-    svm_parameter.C = Cost;
-    svm_parameter.nr_weight = 2;
-    svm_parameter.weight_label = (int *)malloc(2 * sizeof(int)); // deleted in svm_destroy_param
-    svm_parameter.weight_label[0] = -1;
-    svm_parameter.weight_label[1] = 1;
-    svm_parameter.weight = (double *)malloc(2 * sizeof(double));  // deleted in svm_destroy_param
-    svm_parameter.weight[0] = 1;
-    svm_parameter.weight[1] = Unbalance;
-    
-    // cache_size is only needed for kernel functions
-    svm_parameter.cache_size = 100;
-    svm_parameter.eps = 1e-3;
-    //svm_parameter.p = 0.1; // not needed but it has to be positive as it is checked
-    svm_parameter.shrinking = 0;
-    svm_parameter.probability = 0;
-    //  not needed for linear SVM, but it has to be positive as it is checked
-    svm_parameter.gamma = 1.0;
-
-    // load train data
-    svm_problem_t svm_problem;
-    svm_problem.l = VecV.Len();
-    // reserve space for target variable
-    svm_problem.y = (double *)malloc(VecV.Len() * sizeof(double));
-    // reserve space for training vectors
-    svm_problem.x = (svm_node_t **)malloc(VecV.Len() * sizeof(svm_node_t *));
-    // compute number of nonzero elements and get dimensionalit
-    int NonZero = 0, Dim = 0;
-    for (int VecN = 0; VecN < VecV.Len(); ++VecN) {
-        NonZero += (VecV[VecN].Len() + 1);
-        if (!VecV[VecN].Empty()) {
-            Dim = TInt::GetMx(Dim, VecV[VecN].Last().Key + 1);
-        }
-    }
-    svm_node_t* x_space = (svm_node_t *)malloc(NonZero * sizeof(svm_node_t));
-    // load training data and vectors
-    int N = 0, prevN = 0;
-    for (int VecN = 0; VecN < VecV.Len(); ++VecN) {
-        prevN = N;
-        svm_problem.y[VecN] = TargetV[VecN];
-        for (int EltN = 0; EltN < VecV[VecN].Len(); ++EltN) {
-            x_space[N].index = VecV[VecN][EltN].Key+1;
-            x_space[N++].value = VecV[VecN][EltN].Dat;
-        }
-        x_space[N++].index = -1;
-        svm_problem.x[VecN] = &x_space[prevN];
-    }
-
-    const char* error_msg = svm_check_parameter(&svm_problem, &svm_parameter);
-    EAssertR(error_msg == NULL, error_msg);
-
-    // train the model
-    svm_model_t* svm_model = svm_train(&svm_problem, &svm_parameter, DebugNotify, ErrorNotify);
-
-    // compute normal vector from support vectors
-    TFltV WgtV(Dim);
-    TFlt Bias = -svm_model->rho[0]; // LIBSVM does w*x-b, while we do w*x+b; thus the sign flip
-    EAssertR(TLinAlg::Norm(WgtV) == 0.0, "Expected a zero weight vector.");
-    for (int Idx = 0; Idx < svm_model->l; ++Idx) {
-        svm_node_t* SV = svm_model->SV[Idx];
-        while (SV->index != -1) {
-            WgtV[SV->index - 1] += svm_model->sv_coef[0][Idx] * SV->value;
-            ++SV;
+        svm_model_t* svm_model = svm_train(&svm_problem, &svm_parameter, DebugNotify, ErrorNotify);
+        
+        TFltVV SVs(svm_model->l, DimN);
```

Comment:
Same comment as above, no need for separate variables, just use members directly.